src/html.ml

```diff
@@ -77,128 +101,215 @@ let text_of_inline ast =
-            Buffer.add_string b (htmlentities title);
-            Buffer.add_string b "\""
-        end;
-        Buffer.add_string b " />"
-    | Soft_break ->
-        Buffer.add_string b "\n"
-  in
-  loop md
+let start_print p b mds =
+  List.iter (function
+    | Link_def _ -> ()
+    | md ->
+        p.block p b md;
+        Buffer.add_char b '\n'
+    ) mds;
+  p.finish p b
+
+let finish_print _ _ = ()
+
+let print_concat p b l =
+  List.iter (p.inline p b) l
+
+let print_text _ b t =
+  Buffer.add_string b (htmlentities t)
+
+let print_emph p b = function
+  | Normal, _, md ->
+      Buffer.add_string b "<em>";
+      p.inline p b md;
+      Buffer.add_string b "</em>"
+  | Strong, _, md ->
+      Buffer.add_string b "<strong>";
+      p.inline p b md;
+      Buffer.add_string b "</strong>"
+
+let print_code _ b (_, c) =
+  Buffer.add_string b "<code>";
+  Buffer.add_string b (htmlentities c);
+  Buffer.add_string b "</code>"
+
+let print_hard_break _ b =
+  Buffer.add_string b "<br />\n"
+
+let print_soft_break _ b =
+  Buffer.add_string b "\n"
+
+let print_html _ b body =
+  Buffer.add_string b body
+
+let print_url p b label destination title =
+  Buffer.add_string b "<a href=\"";
+  Buffer.add_string b (percent_encode destination);
+  Buffer.add_string b "\"";
+  begin match title with
+  | None -> ()
+  | Some title ->
+      Buffer.add_string b " title=\"";
+      Buffer.add_string b (htmlentities title);
+      Buffer.add_string b "\""
+  end;
+  Buffer.add_string b ">";
+  p.inline p b label;
+  Buffer.add_string b "</a>"
+
+let print_img b label destination title =
+  Buffer.add_string b "<img src=\"";
+  Buffer.add_string b (percent_encode destination);
+  Buffer.add_string b "\" alt=\"";
+  Buffer.add_string b (htmlentities (text_of_inline label));
+  Buffer.add_string b "\"";
+  begin match title with
+  | None -> ()
+  | Some title ->
+      Buffer.add_string b " title=\"";
+      Buffer.add_string b (htmlentities title);
+      Buffer.add_string b "\""
+  end;
+  Buffer.add_string b " />"
 
-let to_html b md =
+let print_link p b = function
+  | Url, {label; destination; title} ->
+    print_url p b label destination title
+  | Img, {label; destination; title}  ->
+    print_img b label destination title
+
+let print_ref p b = function
+  | Url, label, {destination; title; _} ->
+    print_url p b label destination title
+  | Img, label, {destination; title; _} ->
+    print_img b label destination title
+
+let print_inline p b = function
+  | Concat l ->
+    p.concat p b l
+  | Text t ->
+    (* Format.eprintf "Text: %s\n" (id_of_string t); *)
+    p.text p b t
+  | Emph (kind, style, md) ->
+    p.emph p b (kind, style, md)
+  | Code (i, c) ->
+    p.code p b (i, c)
+  | Hard_break ->
+    p.hard_break p b
+  | Soft_break ->
+    p.soft_break p b
+  | Html body ->
+    p.html p b body
+  | Link (kind, link) ->
+    p.link p b (kind, link)
+  | Ref (kind, md, link) ->
+    p.ref p b (kind, md, link)
```

Comment:
It looks like you are not using `ocp-indent`.

Reply:
Yes my bad, will do it now

Reply:
Do you want me to use ocaml-format as well ?

Reply:
No, I haven't looked into it yet. Let's leave that for later.